### ISSUE TITLE
Baseline PWA Support

### ISSUE THREAD

**Issue by phillipskevin** [donejs/donejs#734]
__tldr;__ create a new generator, available via `donejs add pwa` that makes your app into a PWA.

> This was discussed on a [recent live stream](https://youtu.be/wUhv5zIUBv0?t=2518) and previously on [another (54:15)](https://youtu.be/ieZJrPwSZYQ?t=54m15s) and at a [contributors meeting (48:08)](https://youtu.be/95OX6-qii24?t=48m08s)

# The Problem

To make a mobile web app that is competitive with native apps you need a few basic things:

* The ability to add an app icon to the user's home screen.
  * The icon should be high resolution.
* Basic usage of the app when offline.
* A fast loading site.

Luckily DoneJS is way ahead of the curve on provide a fast loading site. For the others, we need to add a new generator.

Google and Firefox together built a few new specifications that are collectively known under the __Progressive Web Apps__ moniker. 

For *baseline* PWA support you need:

* Your app to be served over HTTPS (now supported in done-serve).
* Responsive design for tablet/mobile.
* A [service worker](https://developers.google.com/web/fundamentals/getting-started/primers/service-workers) to make your app available when offline.
* A [web app manifest](https://developers.google.com/web/fundamentals/engage-and-retain/web-app-manifest/) that provides the metadata needed to give your application a home screen icon and a splash page while loading.

# The Solution

Create a new generator, __donejs-pwa__, that will add the necessary files to make the app PWA compliant. It should:

* Update your *build.js* file to generate a *serviceworker.js* script every time you build. This script will contain the code that will cache everything in your `dist/` folder using the [Cache API](https://developer.mozilla.org/en-US/docs/Web/API/Cache), making those assets available when offline.
* Create a *manifest.json* file, this is the web app manifest.
* Update your *index.stache* to load the web app manifest.

@pYr0x did most of the hard part of making service workers supported as part of the steal build process when he created [steal-serviceworker](https://github.com/stealjs/steal-serviceworker). We'll utilize that.

# Future improvements

As this only gives *baseline* PWA support it doesn't help you with everything you need. In particular:

* Data is *not* cached in the service worker. If you are using the can-connect localstorage behavior it might be cached there.

Even if the data *is* cached this won't help with things like form submissions. You will need to add the appropriate functionality to your app yourself. We can help a few different ways:

* Provide standard attributes that make it easy to adjust CSS for when offline. For example if the HTML looks like:

   ```html
   <html offline>
     ...
   </html>
   ```

  Then users can do things like disable buttons:

  ```css
  html[offline] button {
    opacity: 0.65;
    cursor: not-allowed;
  }
  ```
* can-connect behaviors for dealing with offline. When offline, POST and PUT requests could be stored in a queue, and then replayed when the user is back online again.

**Comment by leoj3n** [comment]
https://developers.google.com/web/updates/2017/02/navigation-preload

**Comment by marshallswain** [comment]
Just for kicks, and inspired in part by [Matthew's HTTP 2 article](https://www.bitovi.com/blog/utilizing-http2-push-in-a-single-page-application), I tried running our default `donejs add app` code through the [Chrome Lighthouse extension](https://developers.google.com/web/tools/lighthouse/) that rates PWAs.  

The full checklist for creating a PWA is [here](https://developers.google.com/web/tools/lighthouse/).

Below is the output from the plugin.  If we don't have anybody working on getting the default score higher, I'd like to start contributing towards that goal.

![screen shot 2017-06-05 at 8 03 33 pm](https://cloud.githubusercontent.com/assets/128857/26810738/9ba29c18-4a2a-11e7-9df8-394cf02e7974.jpg)
![screen shot 2017-06-05 at 8 03 54 pm](https://cloud.githubusercontent.com/assets/128857/26810743/9e80b8a2-4a2a-11e7-8e1b-4a333de7e398.jpg)
![screen shot 2017-06-05 at 8 04 08 pm](https://cloud.githubusercontent.com/assets/128857/26810748/a12b4176-4a2a-11e7-8e46-aec6159f4663.jpg)

Yes.  The above is in dev mode.  Below is the report after running `npm run build` and `npm start`.

![screen shot 2017-06-06 at 7 34 20 am](https://cloud.githubusercontent.com/assets/128857/26831792/a207e65c-4a8a-11e7-8ed4-a261b6deb467.jpg)
![screen shot 2017-06-06 at 7 35 13 am](https://cloud.githubusercontent.com/assets/128857/26831828/ba08ec7e-4a8a-11e7-9da1-a54e6e693b58.jpg)

**Comment by chasenlehara** [comment]
@marshallswain The core team has talked about this but we haven’t prioritized it yet. Here are my _personal_ thoughts:

1) There’s probably some low-hanging fruit that we can do for all DoneJS apps to improve the score. In a personal project, the only item under Performance that needs to be improved is “Reduce render-blocking stylesheets”
2) For things that are PWA-specific, we should have [donejs add pwa](https://github.com/donejs/donejs-pwa) to create a manifest file, prompt for app colors, etc.
3) We are _so close_ to being able to offer [advanced PWA features](https://github.com/donejs/donejs/issues/735) like offline support; I’m hoping the core team can focus on this later this summer.

I don’t think you would conflict with anything that the core team is doing if you wanted to work on this issue… and it would be so awesome if you did. 🙏

**Comment by marshallswain** [comment]
Thanks @chasenlehara.  I was thinking the same thing about the `donejs add pwa`.  I created the `donejs-pwa` project on GitHub and pushed a DoneJS-generated generator to it.  We can use that as a starting point.

I'll see if I can get the service worker running and will look at getting done-serve to serve up HTTPS by default, similar to how the [`serve-https` package](https://www.npmjs.com/package/serve-https) does it.
